components/vr/webvr_thread.rs

```diff
@@ -0,0 +1,332 @@
+        self.constellation_chan.send(ConstellationMsg::WebVREvent(pipeline_ids.clone(), event)).unwrap();
+    }
+
+    fn schedule_poll_events(&mut self) {
+        if self.service.is_initialized() && !self.polling_events {
+            self.polling_events = true;
+            let webvr_thread = self.sender.clone();
+            let (sender, receiver) = ipc::channel().unwrap();
+
+            thread::Builder::new().name("WebVRPollEvents".into()).spawn(move || {
+                loop {
+                    if webvr_thread.send(WebVRMsg::PollEvents(sender.clone())).is_err() {
+                        // WebVR Thread closed
+                        break;
+                    }
+                    if !receiver.recv().unwrap_or(false) {
+                        // WebVR Thread asked to unschedule this thread
+                        break;
+                    }
+                    thread::sleep(time::Duration::from_millis(500));
+                }
+            }).expect("Thread spawning failed");
+        }
+    }
+}
+
+// Note about WebVRCompositorHandler implementation:
+// Raw pointers are used instead of Arc<Mutex> as a heavy optimization for latency reasons.
+// This also avoids "JS DDoS" attacks: A Second JSContext doing a lot of calls.
+// while the main one is presenting and demands both high framerate and low latency.
+
+pub struct WebVRCompositor(*mut VRDevice);
+pub struct WebVRCompositorHandler {
+    compositors: HashMap<webrender_traits::VRCompositorId, WebVRCompositor>,
+    webvr_thread_receiver: Receiver<Option<WebVRCompositor>>,
+    webvr_thread_sender: Option<IpcSender<WebVRMsg>>
+}
+
+#[allow(unsafe_code)]
+unsafe impl Send for WebVRCompositor {}
+
+pub type WebVRCompositorSender = Sender<Option<WebVRCompositor>>;
+
+impl WebVRCompositorHandler {
+    pub fn new() -> (Box<WebVRCompositorHandler>, WebVRCompositorSender) {
+        let (sender, receiver) = mpsc::channel();
+        let instance = Box::new(WebVRCompositorHandler {
+            compositors: HashMap::new(),
+            webvr_thread_receiver: receiver,
+            webvr_thread_sender: None
+        });
+
+        (instance, sender)
+    }
+}
+
+impl webrender_traits::VRCompositorHandler for WebVRCompositorHandler {
+    #[allow(unsafe_code)]
+    fn handle(&mut self, cmd: webrender_traits::VRCompositorCommand, texture_id: Option<u32>) {
+        match cmd {
+            webrender_traits::VRCompositorCommand::Create(compositor_id) => {
+                self.create_compositor(compositor_id);
+            }
+            webrender_traits::VRCompositorCommand::SyncPoses(compositor_id, near, far, sender) => {
+                if let Some(compositor) = self.compositors.get(&compositor_id) {
+                    let pose = unsafe {
+                        (*compositor.0).sync_poses();
+                        (*compositor.0).synced_frame_data(near, far).to_bytes()
+                    };
+                    let _result = sender.send(Ok(pose));
```

Comment:
The patterns `let _ = sender.send(..)` or `drop(sender.send())` seem more used across the codebase.